Gridding testing

nannau opened the pull request:
Additions to the coordinate `gridding` with climpyrical:
* Handling coordinates and transforming coordinates into correct CRS
* Flattening coordinates into pairwise points
* Finds nearest indices of stations in coordinate grids
* Finds nearest indices of several stations in coordinate grids
* Finds the model value associated with a station, interpolates if no valid model value exists (i.e. station over water)
* Checks inputs of all the above functions
* Tests many new cases for above functions

corviday commented on climpyrical/.coverage:
```diff
@@ -0,0 +1 @@
+!coverage.py: This is a private format, don't read it directly!{"lines":{"/home/nannau/temp/climpyrical/climpyrical_master/climpyrical/climpyrical/__init__.py":[1],"/home/nannau/temp/climpyrical/climpyrical_master/climpyrical/climpyrical/tests/__init__.py":[1],"/home/nannau/temp/climpyrical/climpyrical_master/climpyrical/climpyrical/tests/test_gridding.py":[1,14,15,18,19,20,21,24,25,27,28,30,31,3 ...
```
I don't know what this file is about, but I have serious doubts that paths to your home directory belong in the repository.

nannau replied:
Thanks. That was accidentally added.

nikola-rados commented on climpyrical/tests/test_gridding.py:
```diff
@@ -0,0 +1,301 @@
+xi = np.linspace(-33.8800048828125, 33.8800048828125, 155)
+yi = np.linspace(-28.59999656677246, 28.15999984741211, 130)
+xext, yext = np.tile(xi, yi.size), np.repeat(yi, xi.shape)
+xext_bad, yext_bad = np.repeat(xi, yi.size), np.tile(yi, xi.shape)
```
So in several places in this file you are defining global variables that are used by the methods underneath.  Instead of creating these variables here, it may be better for you to put them into `conftest.py` .  That way you they can just be passed as parameters rather than defined as globals and passed in from `mark.parametrize`.

```
# in conftest.py
@pytest.fixture(scope='function')
def xext_bad():
    # create the variable here
```

nannau replied:
Can you elaborate on why this would be better? Is `conftest.py` a standard practice for testing?

rod-glover replied:
I disagree in this particular case. The way you are using the global variables here is to construct parametrizations of test functions, and that much harder to do using fixtures. So for this particular use, to keep things simple, I wouldn't move these globals into `conftest.py` as fixtures.

To answer your broader question: `conftest.py` is core to `pytest`. It serves quite a range of test configuration, setup/teardown, and related purposes. It's where you put test fixtures, amongst other things, which are then automagically available globally to tests as function parameters. (Nikola's explanation conflates parametrization and fixture, which are not the same thing.)

rod-glover commented on climpyrical/gridding.py:
```diff
@@ -0,0 +1,478 @@
+            ..."
+        )
+
+    return True
```
Addendum: I see that the `True` return value does make the test code a little clearer (sort of), but I think this is just as good:

```

if passing:
    check_validity()
else:
    with pytest.raises(...):
        check_validity()
```

jameshiebert replied:
One alternative (not necessarily advocating for it in this instance), is to name the function `is_valid` and then return `True` or `False`. It reads a little cleaner and is consistent in its behaviour.

rod-glover replied:
It wasn't so much the semantics of the name (about which @jameshiebert is right), but about the usage, which is in this case solely to raise an exception or fall through. (I add this only as explanation for @nannau 's sake.)

rod-glover replied:
And it's a question of where the responsibility to raise an exception (if that's what you're going to do) lives. I think it's not bad in the check function; that centralizes both the check and the construction of the exception (rather, its message).

nannau replied:
The return value is unnecessary, unless I was asking an `is_valid` type question, which I'm not, I'm just checking the values are believable. The purpose is to raise an exception or fall through. 

Is removing the return statement best here? Or are there other semantics that could be improved upon?

rod-glover replied:
Just a learning point: If you were constructing an `is_valid` function, then it should return True or False, and not raise an exception. Alternatively, the equivalent `check` function should only raise an exception.

Unlike many other languages, good Python style favors exception handling: act first, ask permission later, which is to say don't test, just handle the exception.

So I would say that removing the return statement is best here.

nannau replied:
Okay - thank you!

rod-glover commented on climpyrical/tests/test_gridding.py:
```diff
@@ -0,0 +1,301 @@
+    if passed:
+        assert check_input_coords(x, y)
+    else:
+        with pytest.raises((ValueError, TypeError)):
```
Suggest you parametrize which error you expect to be raised in each case.

nannau replied:
I knew there was a better way to do that. Thanks. That's an answer to a question I've been meaning to ask but hasn't come up yet.

nannau commented:
@jameshiebert, @rod-glover and @nikola-rados could I get a LGTM from any of you on this PR when you get a chance?

I've addressed the following suggestions:
* Removed accidental `.coverage` file
* Remove return statement from all `check` functions
* Provide more detailed exception description guiding user on the error raised based on their inputs 
* Added an additional `check_ndims` check function that ensures an array is of the expected size.
* Parameterizes the expected error to be found in the context manager with `pytest.raises`

rod-glover commented on climpyrical/gridding.py:
```diff
@@ -0,0 +1,535 @@
+        ValueError:
+            If data's dimension is not expected
+    """
+    if not isinstance(data, np.ndarray):
```
**This is not a change request**, just a bit of advice: 

Whilst checking the types of all your arguments is admirable, it is also laborious, time consuming, and can be done better with the new-ish [type annotations](https://docs.python.org/3/library/typing.html). If something else would burk (a.k.a. raise an exception) for wrong-type args, then don't bother type checking ...

Unless it is utterly central to your code and there is a distinct possibility of it being fooled. Or unless these functions will be exposed for wide use by others. Then you  want more careful checking and very specific error messages.

nannau replied:
I mainly included them to guarantee that the follow-up attributes I need to check, i.e. `max()`, `min()`, `size`, and some of the functions like `np.any` or `np.isnan` will definitely work with any numpy.narray provided. 

I agree, typing is the way forward. Down the road, look for future PRs that include that. 

I know there are some objects other than `np` that share some of these attributes, but I'm not sure if `climpyrical` should have multiple object compatibility just yet. I do see that as a nice feature down the road.

nannau replied:
Read that as: I'm not sure if `numpy.ndarrays` are even the best data type for `climpyrical` to pass around. I am looking into perhaps using just `xarray` objects or `dask` objects in the future.

rod-glover replied:
That's a much bigger discussion.
